AutoCollection/Statsbeat.ts

```diff
@@ -0,0 +1,263 @@
+        var requests = {
+            totalRequestCount: this._totalRequestCount,
+            time: +new Date
+        };
+        var intervalRequests = (requests.totalRequestCount - lastRequests.totalRequestCount) || 0;
+        var elapsedMs = requests.time - lastRequests.time;
+        var averageRequestExecutionTime = ((this._intervalRequestExecutionTime - this._lastIntervalRequestExecutionTime) / intervalRequests) || 0; // default to 0 in case no requests in this interval
+        this._lastIntervalRequestExecutionTime = this._intervalRequestExecutionTime; // reset
+        if (elapsedMs > 0 && intervalRequests > 0) {
+            this._statbeatMetrics.push({ name: Constants.StatsbeatCounter.REQUEST_DURATION, value: averageRequestExecutionTime });
+        }
+        this._lastRequests = requests;
+    }
+
+    private _trackRequestsCount() {
+        if (this._totalSuccesfulRequestCount > 0) {
+            this._statbeatMetrics.push({ name: Constants.StatsbeatCounter.REQUEST_SUCCESS, value: this._totalSuccesfulRequestCount });
+            this._totalSuccesfulRequestCount = 0; //Reset
+        }
+        if (this._totalFailedRequestCount > 0) {
+            this._statbeatMetrics.push({ name: Constants.StatsbeatCounter.REQUEST_FAILURE, value: this._totalFailedRequestCount });
+            this._totalFailedRequestCount = 0; //Reset
+        }
+        if (this._retryCount > 0) {
+            this._statbeatMetrics.push({ name: Constants.StatsbeatCounter.RETRY_COUNT, value: this._retryCount });
+            this._retryCount = 0; //Reset
+        }
+        if (this._throttleCount > 0) {
+            this._statbeatMetrics.push({ name: Constants.StatsbeatCounter.THROTTLE_COUNT, value: this._throttleCount });
+            this._throttleCount = 0; //Reset
+        }
+        if (this._exceptionCount > 0) {
+            this._statbeatMetrics.push({ name: Constants.StatsbeatCounter.EXCEPTION_COUNT, value: this._exceptionCount });
+            this._exceptionCount = 0; //Reset
+        }
+    }
+
+    private async _sendStatsbeats() {
+        let envelopes: Array<Contracts.Envelope> = [];
+        let properties = {
+            "os": this._os,
+            "rp": this._resourceProvider,
+            "cikey": this._cikey,
+            "runtimeVersion": this._runtimeVersion,
+            "language": this._language,
+            "version": this._sdkVersion,
+            "attach": this._attach,
+            "instrumentation": this._instrumentations,
+            "feature": this._features,
+        }
+        for (let i = 0; i < this._statbeatMetrics.length; i++) {
+            let statsbeat: Contracts.MetricTelemetry = {
+                name: this._statbeatMetrics[i].name,
+                value: this._statbeatMetrics[i].value,
+                properties: properties
+            };
+            let envelope = EnvelopeFactory.createEnvelope(statsbeat, Contracts.TelemetryType.Metric, null, null, this._statsbeatConfig);
```

Comment:
it looks like config is used to store statsbeat connection string only.  can you just pass the connection string to createEnvelop? unless you will parse connection string from the json config file later?

Reply:
We are using same mechanism as other envelopes created in the SDK, I'm trying to avoid specific code paths just for Statbeat here, added connection string as static so it could be updated if we add JSON config later